test/multiplier/txn.go

```diff
@@ -0,0 +1,297 @@
+//     a new store (e.g. [NewStore], [NewBadgerStore]).
+//   - Inserting a [CommitTxn] immediately before the first action that closes the store
+//     (e.g. [CancelCtx], [CloseStore]).
+//   - Copying all original read actions after the last write action and moving the copies
+//     to immediately after the inserted [CommitTxn] action.
+//   - Modifying all original read and write actions (e.g. [GetValue], [SetValue]) to apply to
+//     the created transaction instead of the underlying store.
+type txnCommit struct{}
+
+var _ Multiplier = (*txnCommit)(nil)
+
+func (n *txnCommit) Name() Name {
+	return TxnCommit
+}
+
+func (n *txnCommit) Apply(source action.Actions) action.Actions {
+	lastCreateStoreIndex := 0
+	firstCloseIndex := 0
+	newActions := make(action.Actions, 0, len(source))
+	postCommitActions := action.Actions{}
+
+	for i, a := range source {
+		switch a.(type) {
+		case *action.NewStore, *action.NewBadgerStore, *action.NewMemoryStore:
+			lastCreateStoreIndex = i
+
+		case *action.CancelCtx, *action.CloseStore:
+			if firstCloseIndex == 0 {
+				firstCloseIndex = i
+			}
+
+		case *action.CreateNewTxn:
+			// If the action set already contains txns we should not adjust it
+			return source
+		}
+	}
+
+	for i, a := range source {
+		switch a.(type) {
+		case *action.NamespaceStore:
+			newActions = append(newActions, a)
+			newActions = append(newActions, action.WithTxn(a))
+			if i < firstCloseIndex {
+				firstCloseIndex += 1
+			}
+			if i < lastCreateStoreIndex {
+				lastCreateStoreIndex += 1
+			}
+			continue
+
+		case *action.NewStore, *action.NewBadgerStore, *action.NewMemoryStore:
+			newActions = append(newActions, a)
+			continue
+
+		case *action.CancelCtx, *action.CloseStore:
+			newActions = append(newActions, a)
+			continue
+
+		case *action.GetValue, *action.HasValue, *action.Iterate, *action.Iterator:
+			if firstCloseIndex == 0 {
+				postCommitActions = append(postCommitActions, a)
+			}
+
+		case *action.DeleteValue, *action.SetValue:
+			postCommitActions = action.Actions{}
+		}
+
+		if i > lastCreateStoreIndex && firstCloseIndex == 0 {
+			newActions = append(newActions, action.WithTxn(a))
+		} else {
+			newActions = append(newActions, a)
+		}
+	}
+
+	// Add two to the length in order to accomodate the new txn and discard actions
+	result := make(action.Actions, len(newActions)+2+len(postCommitActions))
+
+	indexOffset := 0
+	for i, a := range newActions {
+		newIndex := func() int {
+			return i + indexOffset
+		}
+
+		if i == lastCreateStoreIndex {
+			result[newIndex()] = a
+
+			indexOffset += 1
+			result[newIndex()] = action.NewTxn()
+		} else if i == firstCloseIndex {
+			result[newIndex()] = action.Commit()
+			indexOffset += 1
+
+			for _, postCommitAction := range postCommitActions {
+				result[newIndex()] = postCommitAction
+				indexOffset += 1
+			}
+
+			result[newIndex()] = a
+		} else {
+			result[newIndex()] = a
+		}
+	}
```

Comment:
suggestion: You could simplify this with a similar solution as well.

Reply:
I am even more sceptical as to whether this would be simpler using a similar solution as proposed earlier, this loop is a bit more involved.  Leaving as-is for now.